fix: aligning and distributing elements and nested groups while editing a group

cTangonan123 opened the pull request:
## Summary
This PR fixes a bug related to the align and distribute actions that prevented their usage while in edit group mode. Previously, when editing the contents of a group, align/distribute actions were disabled for selected elements and/or nested groups.
- both align and distribute actions had a similar structure in organizing selectedElements into groups, although they both seemed to use `getMaximumGroups()` which is also used in `snapping.ts` and used in order to group existing elements in the scene that are not selected.

https://github.com/user-attachments/assets/57975b69-4849-4a9f-8139-f55102bff86c

### Fix
The logic has been updated to ensure that align and distribute operations work as expected within the context of group editing and can still pass all previously written tests for the align action.
- addition of a new method `getSelectedElementsByGroup` which is actually very similar to `getMaximumGroups` but returns the maximum selected grouping(could be renamed to reflect that).
- this new method is also used in `actionAlign.ts` and `actionDistribute.ts` in their predicates to show only when the tool can be used.

#### Feedback Welcome
I'm happy to adjust anything as needed—please let me know if there's anything that could be improved or if there's a better way to approach the fix. Appreciate any suggestions or review comments!

### Related issue
Fixes #9606 

cc @Mrazator  — appreciate your input on this! Let me know if it resolves what you noticed, and I'm happy to adjust anything if needed.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

* **New Features**
  * Improved alignment and distribution actions to better respect element grouping and nested group structures.
  * Enhanced logic for enabling alignment and distribution actions based on grouped selection, providing more intuitive behavior in group edit modes.

* **Bug Fixes**
  * Alignment and distribution now work correctly with elements inside groups and nested groups.

* **Tests**
  * Added comprehensive tests covering alignment behaviors for grouped and nested group scenarios, including group edit mode.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Mrazator reviewed (commented):
First of all, thanks for the PR!

As shown in the recording, that flow with multiple selected elements & groups seems to work as expected. 

Additionally, I think it would be useful to make the align / distribute actions work when selecting a single group (and possibly selected group within a group + multiple selected groups within a group), without having to select individual items inside.

The basic flow is recorded here: https://github.com/excalidraw/excalidraw/issues/9606#issuecomment-2963788008

cTangonan123 commented on packages/element/src/groups.ts:
```diff
@@ -402,3 +402,31 @@ export const getNewGroupIdsForDuplication = (
+  selectedElements.forEach((element: ExcalidrawElement) => {
+    const groupId =
+      getSelectedGroupIdForElement(element, appState.selectedGroupIds) ||
+      element.id;
```
I definitely see where confusion could happen, thanks for addressing....would you consider syntax change to the variable `group`, instead naming it something like `containers`?

Mrazator replied:
We have `containers` keyword reserved for text containers (elements which have bound text elements), maybe just have two separate functions and refer to each as `groups` and `elements`.

cTangonan123 commented:
> As shown in the recording, that flow with multiple selected elements & groups seems to work as expected.
> 
> Additionally, I think it would be useful to make the align / distribute actions work when selecting a single group (and possibly selected group within a group + multiple selected groups within a group), without having to select individual items inside.

I like it @Mrazator ! and think I'm capable, from looking at the apps mentioned in [#9606(comment)](https://github.com/excalidraw/excalidraw/issues/9606#issuecomment-3034418743) their case for aligning of a single group selection is to the artboard, but since Excalidraw has infinite canvas it leaves open the opportunity to differentiate. 

I think It's possible for me to set it up conditionally within the new method, just wanting confirmation on the situation below:

https://github.com/user-attachments/assets/b018e8d7-dba4-4c36-9c54-362d26dbf4c0

to my understanding, from what was mentioned all elements within the group selected would align to the left correct?

cTangonan123 commented on packages/excalidraw/actions/actionAlign.tsx:
```diff
@@ -38,7 +40,11 @@ export const alignActionsPredicate = (
+      selectedElements,
+      app.scene.getNonDeletedElementsMap(),
+      appState as Readonly<AppState>,
+    ).length > 1 &&
```
If we go the route I'm currently thinking of altering `getSelectedElementsByGroup` to check for the case in which a single `groupId` is selected and nothing else, I think it could work as the length instead would count for every element within the group...open to any suggestions you may have on this.

cTangonan123 commented:
> Yes, but this case is a bit special, since the last group you made is the top-most group, which contains the last element and a group of the first three elements. Therefore, "all elements within the group" is not precise, since it's one group and one element. And since the group is already on the left, only the last element would be aligned to the left.

@Mrazator totally understood now thanks for clarifying! With respect to the hierarchy 👑, That makes it a bit more of a challenge but am up for it!
I think it could possibly be determinant of what indices `selectedGroupId` is in accordance to each of the `selectedElements` `groupIds`. Will continue to work on it and tag you if I make any progress.

cTangonan123 commented:
hey @Mrazator !
Got everything working for all the cases, The changes include:
- added an `elements` map to distinguish between groups and elements
- Introduced helper functions: `addToElements`, `addToGroups`, and `getKeyIdForNestedElement`

It took a few iterations to refactor, but your suggestion to separate out the logic into functions really helped clarify the flow –much appreciated! 🙏

https://github.com/user-attachments/assets/36e66572-3e76-4914-8feb-0e01d2a3dbf2

Would love any additional feedback you may have when you get a chance. Thanks again for your time!

Mrazator reviewed (approved):
Great work! I don't have much to comment about, other than some syntactical nitpicks.

Especially appreciated adding the tests for all the different selection / align cases.

cTangonan123 commented:
Awesome @Mrazator ! I'll go ahead and map out some case/conditions today to test for and send them your way to check for any edge cases I might miss. `distribute.ts` was a bit intimidating to learn to be quite honest lol.
